### ISSUE TITLE
How did u train sfnet_resnet_dsn??

### ISSUE THREAD

**Issue by jinfagang** [lxtGH/SFSegNets#35]
Please have a close look at  your code:

```
def forward(self, x, gts=None):
        x_size = x.size()  # 800
        x0 = self.layer0(x)  # 400
        x1 = self.layer1(x0)  # 400
        x2 = self.layer2(x1)  # 100
        x3 = self.layer3(x2)  # 100
        x4 = self.layer4(x3)  # 100
        x = self.head([x1, x2, x3, x4])
        # main_out = Upsample(x[0], x_size[2:])
        main_out = F.upsample(x[0], size=x_size[2:], mode='bilinear', align_corners=True)
        # main_out = Upsample(x[0], x_size[2:])
        print('main out: {}'.format(main_out))
        print(main_out.shape)
        if self.training:
            if not self.fpn_dsn:
                return self.criterion(main_out, gts)
            return self.criterion(x, gts)
        return main_out
```

in sfnet_resnet line 178, if not self.fpn_dsn which I set dsn to be True, then:

You will using `return self.criterion(x, gts)`
But, x is a tuple.....
How do u able to train in this logic???

CrossEntropy2D can not performance .size or .dim on tuple, do u know?

I am totally don't know how do u able to train it.

**Comment by lxtGH** [comment]
Still don't know what you mean. 
I try my default configs. 
![image](https://user-images.githubusercontent.com/17202268/113993916-80922380-9887-11eb-9aea-521c6758c931.png)

![image](https://user-images.githubusercontent.com/17202268/113993945-8851c800-9887-11eb-9268-d53e8d1c3bd2.png)

with fpn_dsn and without fpn_dsn. Both work well.

**Comment by jinfagang** [comment]
@lxtGH Please take a deep look, if using fpn_dsn, you will got to:

```
if self.training:
            if not self.fpn_dsn:
                return self.criterion(main_out, gts)
            return self.criterion(x, gts)
```

You will call `return self.criterion(x, gts)`, and ** did u see what is x?**. Its output from your head which is a dict.

Your criterion doesn't supports dict as inputs.

BTW, my pytorch version is 1.7
